content/python_files/classification_metrics.py

```diff
@@ -0,0 +1,219 @@
+
+y_observed = np.asarray([1, 0, 1, 1, 0])
+y_predicted_probs = np.asarray([0.1, 0.9, 0.3, 0.5, 0.2])
+
+
+def compare_metrics(metric_func, transformation_func, y_observed, y_predicted_probs):
+    metric_name = metric_func.__name__
+    a = metric_func(y_observed, y_predicted_probs)
+    b = metric_func(y_observed, transformation_func(y_predicted_probs))
+
+    print(f"{metric_name} on original soft predictions: {a:.4f}")
+    print(f"{metric_name} on transformed soft predictions: {b:.4f}")
+
+
+compare_metrics(roc_auc_score, monotonic_function, y_observed, y_predicted_probs)
+
+# %% [markdown]
+#
+# ### Solution
+#
+# - Check that the same result holds for other ranking metrics such as
+# `average_precision_score`.
+#
+# - Tweak the values in `y_predicted_probs` to see that this
+# property holds in general.
+
+
+# %%
+from sklearn.metrics import average_precision_score
+
+
+# TODO: tweak me!
+y_predicted_probs = np.asarray([0.1, 0.9, 0.3, 0.5, 0.2])
+
+# TODO: write me!
+
+# %% [markdown]
+# ### Solution:
+
+# %%
+from sklearn.metrics import average_precision_score
+
+compare_metrics(
+    average_precision_score, monotonic_function, y_observed, y_predicted_probs
+)
+
+# %%
+y_predicted_probs = np.asarray([0.4, 0.9, 0.3, 0.5, 0.0])
+compare_metrics(roc_auc_score, monotonic_function, y_observed, y_predicted_probs)
+compare_metrics(
+    average_precision_score, monotonic_function, y_observed, y_predicted_probs
+)
+
+# %% [markdown]
+# ## Proper scoring rules are impacted by monotonic transformations
+#
+# ### Exercise:
+#
+# - Check that `neg_log_loss` and `brier_score_loss` are impacted by monotonic
+#   transformations.
+
+# %%
+
+
+# TODO: write me!
+
+
+# %% [markdown]
+# ### Solution:
+
+# %%
+from sklearn.metrics import log_loss, brier_score_loss
+
+compare_metrics(log_loss, monotonic_function, y_observed, y_predicted_probs)
+
+# %%
+compare_metrics(brier_score_loss, monotonic_function, y_observed, y_predicted_probs)
+
+# %% [markdown]
+#
+# There is no particular reason to expect that our choice of transformation
+# would improve calibration on this dataset. Ranking is unchanged (as measured
+# by ROC-AUC). As a result, the proper scoring losses are degraded.
+
+# %% [markdown]
+#
+# ## Hard classification metrics only depends on the value of the threshold
+#
+# ### Question:
+#
+# - Under which conditions are thresholded / hard classification metrics such
+#   as (accuracy, precision, recall, F1 score, ...) impacted by monotonic
+#   transformations of the predicted probabilities?
+
+# %%
+# TODO: write me!
+
+# %% [markdown]
+# ### Solution:
+#
+# - Thresholded metrics are not impacted by monotonic transformations if the
+#   threshold is left unchanged at 0.5 after the transformation.
+
+
+# %%
+def threshold_preserving_transformation(x):
+    return expit((x - 0.5) * 10)
+
+
+x = np.linspace(0, 1, 1000)
+y = threshold_preserving_transformation(x)
+_ = plt.plot(x, y)
+_ = plt.hlines(0.5, 0, 1, colors="gray", linestyles="--")
```

Comment:
Usually, I'm only catching the last line only. But use the style that you prefer and I'll make the rest of the notebooks consistent with it.

```suggestion
plt.plot(x, y)
plt.hlines(0.5, 0, 1, colors="gray", linestyles="--")
```